templates/todo/api/java/src/main/java/com/microsoft/azure/simpletodo/configuration/AzureDeveloperCliCredential.java

```diff
@@ -0,0 +1,220 @@
+import com.azure.core.credential.TokenRequestContext;
+import com.azure.core.exception.ClientAuthenticationException;
+import com.azure.core.util.CoreUtils;
+import com.azure.core.util.logging.ClientLogger;
+import com.azure.core.util.serializer.JacksonAdapter;
+import com.azure.core.util.serializer.SerializerAdapter;
+import com.azure.core.util.serializer.SerializerEncoding;
+import com.azure.identity.CredentialUnavailableException;
+import com.azure.identity.implementation.IdentityClientOptions;
+import com.azure.identity.implementation.util.LoggingUtil;
+import com.azure.identity.implementation.util.ScopeUtil;
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.nio.charset.StandardCharsets;
+import java.time.LocalDateTime;
+import java.time.OffsetDateTime;
+import java.time.ZoneId;
+import java.time.ZoneOffset;
+import java.time.format.DateTimeFormatter;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import java.util.regex.Pattern;
+import reactor.core.publisher.Mono;
+
+/**
+ * A credential provider that provides token credentials based on
+ * Azure Developer CLI command.
+ */
+@Immutable
+public class AzureDeveloperCliCredential implements TokenCredential {
+
+    private static final ClientLogger LOGGER = new ClientLogger(AzureDeveloperCliCredential.class);
+    private IdentityClientOptions identityClientOptions;
+
+    private static final String WINDOWS_STARTER = "cmd.exe";
+    private static final String LINUX_MAC_STARTER = "/bin/sh";
+    private static final String WINDOWS_SWITCHER = "/c";
+    private static final String LINUX_MAC_SWITCHER = "-c";
+    private static final String DEFAULT_WINDOWS_SYSTEM_ROOT = System.getenv("SystemRoot");
+    private static final String DEFAULT_MAC_LINUX_PATH = "/bin/";
+    private static final String WINDOWS_PROCESS_ERROR_MESSAGE = "'azd' is not recognized";
+    private static final Pattern LINUX_MAC_PROCESS_ERROR_MESSAGE = Pattern.compile("(.*)azd:(.*)not found");
+    private static final Pattern ACCESS_TOKEN_PATTERN = Pattern.compile("\"accessToken\": \"(.*?)(\"|$)");
+    private static final SerializerAdapter SERIALIZER_ADAPTER = JacksonAdapter.createDefaultSerializerAdapter();
+
+    /**
+     * Creates an AzureDeveloperCliCredential with default identity client options.
+     */
+    public AzureDeveloperCliCredential() {
+        identityClientOptions = new IdentityClientOptions();
+    }
+
+    @Override
+    public Mono<AccessToken> getToken(TokenRequestContext request) {
+        return authenticateWithAzureCli(request)
+            .doOnNext(token -> LoggingUtil.logTokenSuccess(LOGGER, request))
+            .doOnError(error -> LoggingUtil.logTokenError(LOGGER, identityClientOptions, request, error));
+    }
+
+    private Mono<AccessToken> authenticateWithAzureCli(TokenRequestContext request) {
+        StringBuilder azCommand = new StringBuilder("azd auth token --output json --scope ");
```

Comment:
```suggestion
        StringBuilder azCommand = new StringBuilder("azd auth token --output json");
```

Reply:
I think without this we end up with something that looks like:

```
azd auth token --output json --scope --scope <some-scope>
```

which we don't want.

Reply:
only if there's not a scope in the request (which I don't think will ever happen).

the next line `var scopes = String.join(" --scope ", request.getScopes());` uses `join`, so, if there's only one scope, it would be appended to the end of the command as `azd auth token --output json --scope scopeName`.
If there are more than one, `join` will add the `--scope ` before the second scope

Reply:
Yeah - my bad on understanding what the code was doing.

Reply:
That definitely tripped me up until I realized `" --scope"` was the separator string..

Reply:
@vhvb1989 For defensiveness, maybe we should validate non-empty? I'd hate for a cryptic error message "--scope parameter is missing" to show up somewhere :D

Reply:
that's handed by `ScopeUtil.validateScope()` which throws `IllegalArgumentException` and makes us return `return Mono.error(LOGGER.logExceptionAsError(ex));`